src/piton/labelers/core.py

```diff
@@ -0,0 +1,381 @@
+                X-----(X+A)------(X+B)------
+            
+            
+            This has a label of TRUE:
+                X-----(X+A)--O---(X+B)------
+            
+            This has a label of TRUE:
+                X-----(X+A)--O---(X+B)----O-
+            
+            This has a label of FALSE:
+                X---O-(X+A)------(X+B)------
+            
+            This has a label of FALSE:
+                X-----(X+A)------(X+B)--O---
+        """
+        pass
+
+    def get_prediction_times(self, patient: Patient) -> List[datetime.datetime]:
+        """
+        Return a sorted list containing the datetimes for which it's valid to make a prediction,
+        
+        IMPORTANT: Must be sorted ascending (i.e. start -> end of timeline)
+        """
+        return [ e.start for e in patient.events ]
+
+    def label(self, patient: Patient) -> List[Label]:
+        """
+        Assumes that events in `patient.events` are already sorted in chronologically 
+        ascending order (i.e. start -> end).
+
+        Args:
+            patient (Patient): A patient object
+
+        Returns:
+            List[Label]: A list containing a label for each datetime returned by `get_prediction_times()`
+        """
+        if len(patient.events) == 0:
+            return []
+
+        __, end_time = self.get_patient_start_end_times(patient)
+        outcome_times: List[datetime.datetime] = self.get_outcome_times(patient)
+        time_horizon: TimeHorizon = self.get_time_horizon()
+        
+        results: List[Label] = []
+        curr_outcome_idx: int = 0
+        # For each prediction time, check if there is an outcome which occurs within the (start, end) of the time horizon
+        for time in self.get_prediction_times(patient):
+            while (
+                curr_outcome_idx < len(outcome_times)
+                and outcome_times[curr_outcome_idx] <= time
+            ):
+                # This is the idx in `outcome_times` that corresponds to the first outcome EQUAL or AFTER the prediction time (if one exists)
+                curr_outcome_idx += 1
+
+            # TRUE if an event occurs within the time horizon
+            is_outcome_occurs_in_time_horizon: bool = (
+                curr_outcome_idx < len(outcome_times)
+                and (time + time_horizon.start <= outcome_times[curr_outcome_idx] <= time + time_horizon.end)
+            )
+            # TRUE if patient is censored (i.e. timeline ends BEFORE this time horizon ends,
+            # so we don't know if the outcome happened after the patient timeline ends)
+            is_censored: bool = end_time < time + time_horizon.end
+
+            if is_outcome_occurs_in_time_horizon:
+                results.append(Label(time=time, value=True, label_type="boolean"))
+            elif not is_censored:
+                # Not censored + no outcome => FALSE
+                results.append(Label(time=time, value=False, label_type="boolean"))
```

Comment:
This logic is wrong when time_horizon.start != 0.

I think it can be fixed by changing the initial "while" loop to 

` and outcome_times[curr_outcome_idx] < time + time_horizon.start`

but this needs really extensive testing.